README
This is a program to play Blue Moon against an AI.  Rules for the game
can be found at:

http://www.fantasyflightgames.com/PDF/bluemoonrules.pdf

COMPILATION

As of version 1.0.3, autoconf support has been added.  Hopefully,
compilation should be the typical "./configure && make && make install".
If you experience compilation errors, please send details of your
system and any errors to keldon@keldon.net .

RUNNING

Run with "./bluemoon".  You can specify which People to play and to play
against with the following options:

	-1 <people>   -- People you will play
	-2 <people>   -- People you will play against

You may also change these setting once the game is running with the
"Select Matchup" option under the Game menu.

INTERFACE

To play a card, click on it.  Cards that cannot be legally played are
displayed in monochrome.  If you have a retrievable card on the table,
you may retrieve it by clicking on it.  If you have a card with text
that begins "On my turn", click on it to use its special text effect.

If you wish to play a card that says "I may", and you want to not use
its special effect, click on it with your second, third, etc mouse
button.  This is typically your middle or right mouse button.   This
will bring up a menu with an option to "Play with no effect".  Note that
this feature does not work on several cards at the moment, including all
cards where the special effect takes place at a future point ("at the
end of my turn", for example).

You can also use this popup menu to play cards as bluff cards or to
load them onto a Buka ship.

If an opponent plays a Caterpillar or similar card and you wish to
discard to satisfy it, click on the opponent's card.  Cards like
"Kabu Kat" and "Baku But" will be satisified automatically when you
play a support or booster card, respectively.

Once you are done playing cards, choose to announce your power in Fire
or Earth with the buttons on the left side.  The button will not be
clickable unless you have matched your opponent's power, or played a
shield, etc.  The button will also be unclickable if your opponent
has an unsatisfied Caterpillar card.

If you make a mistake and are not able to match your opponent's Power,
you may undo your turn back to the "Retreat from Fight" phase with the
Undo button.  You may then either retreat or play different cards in
an attempt to match power.

BUGS

If you encounter a bug, especially an error in the rules implementation,
please report it to keldon@keldon.net .

LEGAL

The source code is my own creation and is placed under the GPL.  For
details, see the file COPYING.

All copyrights related to Blue Moon are owned by Kosmos.  Blue Moon was
designed by Reiner Knizia.  Permission to distribute the game, including
the card images, has been granted by Kosmos.

http://www.kosmos.de/
http://www.knizia.de/


ChangeLog
2008-07-27

Released version 1.3.2

2008-07-27

Fixed Boatswain Mena Marn.

2008-07-23

Fix cards which ignore all icons or special text to include influence
cards and any other types of cards.

2008-07-22

Automatically select target for "one of my" cards to maximize total power.

2008-07-21

Fixed GUI turn order when human player bluff is called.

Fixed STOP icon and cards that allow additional booster/support cards.

2008-07-20

Changed Mimix mutant to not count ships.

Adjusted height of cards loaded on ships to avoid running off the bottom.

2008-07-18

Released version 1.3.1

2008-07-17

Fix bug in GUI when choosing from multiple bluff cards.

2008-07-14

Modify README with instructions for playing bluff cards or loading cards
onto ships.

Add code for choosing cards to discard when opponent's hand is disclosed
and cards are discarded from hand.

2008-07-09

Fixed timing of special effects that affect "one of my" cards.

Fixed Lookout Dolora Paal when some support cards were PROTECTED.

Allow leadership or influence cards to be played from the hand even when
a ship is landed.

Always prompt user for target of "ony of my" cards whenever needed,
instead of waiting for announce power.

Split dragons attracted from retreat from dragon from Irresistible Sirens
when opponent is forced to retreat due to Floods.

2008-07-08

Released version 1.3.0

2008-07-08

Added strings to assist feature for Buka related actions such as landing
or loading ships, or playing bluff cards.

2008-07-07

Clear "played this turn" flag on ships when player retreats.

2008-06-12

Fixed bug regarding Old Goo Laa's Ghost and landed ships.

2008-06-11

Performance enhanced when AI considers playing Quartermaster Kura Ko
and Armory in the same turn.

2008-05-29

Improved performance of checking whether a set of cards used to satisfy
a Caterpillar are legal.

2008-05-28

Changed neural net inputs to count total number of cards in hand plus
cards in draw deck, rather than just draw deck.  Otherwise cards like
Muster Reinforcements are sometimes thought to be harmful to play.

2008-05-26

Added code to further predict the consequences of declining a fight.

2008-05-24

Added code to predict whether the opponent will be forced to retreat due
to our play.  Uses cards in hand if they are all known (due to end of
game or disclosure), otherwise uses all cards with unknown locations.

2008-05-01

Reordered "end of support" phase events so that possible game-ending cards
like Irresistible Sirens are handled properly.

2008-04-27

Prevent ignored bluff cards from being called.

Prevent play of bluff cards when support cards are ignored.

2008-04-25

Fix AI bug caused by instant win/loss from Irresistible Sirens when pending
card choice is to be made.

2008-04-05

Added code to handle cards like "Cast Cataclysm" discarding bluff cards.

2008-02-28

Added several neural net inputs for cards loaded on ships, and numbers of
bluff cards played.

Changed in memory structure of neural net for better cache performance
with extra (often unused) inputs.

2008-02-23

Added code to GUI to reveal bluff cards at beginning of turn.

2008-02-20

Fixed bug where cards with values ignored could satisfy Caterpillar
Fireeater or Caterpillar Eartheroder.

Fixed bug where cards could be added to gangs even after card with STOP
icon had been played.

2008-02-18

Added popup menu to cards in hand, to allow loading onto ships or
playing as a bluff.

2008-02-17

Don't allow AI to consider playing cards that were drawn randomly in
simulated games.

Reset random seed in simulated games, so that they don't accurately
predict cards drawn from deck.

2008-02-14

Added GUI code to display influence cards, and cards loaded onto ships
in influence area.

2008-02-13

Removed "upside down" effect in opponent cards, and reuse more code
in table drawing.

2008-02-08

Added code for calling bluffs, and AI code to make bluff calling
decisions.

2008-02-07

Added code for Buka bluff cards.

2008-02-06

Modified AI card chooser routines to work in more difficult situations.

2008-02-05

Added code for Buka ships.

2008-02-04

Fixed bug that allowed cards played "as FREE" to work even against Mad
Mike Magpie.

Added Buka cards to cards.txt file.

Began adding code for Buka card effects.

2008-02-03

Reset full-size image to card back when matchup is changed.

2008-02-01

Removed attempt to hid close button on dialog boxes.  It does not work
well under Windows.

Added extra inputs for "instant win/loss" and train the AI more
strictly when this happens.

2008-01-31

Released version 1.1.1

2008-01-31

Added comments to "do_matchup" script.

Changed to set of newly trained neural nets.

2008-01-30

Cleared text log before loading new neural nets when matchup chosen,
so that error messages would be seen.

Wall of Fire incorrectly prevented searching your discard deck and
taking cards from it.

Added crude "Assist" feature to ask AI about the best move.

2008-01-28

Set random event flag when hand is disclosed, so that AI can recheck
options with the new information.

2008-01-27

Fixed display bug where Administer Water of Immortality didn't clear
the "top of discard pile" card displayed.

Made several changes to "condition" program.  Added training of the
"game over" and "no cards" inputs, and fixed the computed score when
dragons are on one player's side or the other.

2008-01-24

Changed undo button to always go back to beginning of turn, instead of
"retreat?" phase.

Fixed announced power when Firewater or Earthwater is played on the first
turn of a fight.

Fixed memory leak in GUI, where rotated card images were not deleted.

2008-01-23

Added -r flag for setting random seed to learner.

Somewhat unified the AI "move selection" and "card choice" code to
avoid duplicating work.

Removed ugly cache code from net.c, as we now make many fewer duplicate
calls to eval_game(), so the cache code is actually a slowdown now.


init.c
/*
 * Bluemoon AI
 * 
 * Copyright (C) 2007-2008 Keldon Jones
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include "bluemoon.h"

/*
 * Set of peoples.
 */
people peoples[MAX_PEOPLE];


/*
 * Names of icon flags.
 */
static char *icon_name[15] =
{
	"SHIELD_FIRE",
	"SHIELD_EARTH",
	"STOP",
	"RETRIEVE",
	"PAIR",
	"FREE",
	"PROTECTED",
	"GANG_COOL",
	"GANG_TOP",
	"GANG_FUN",
	"GANG_NO",
	"BLUFF_FIRE",
	"BLUFF_EARTH",
	"BLUFF_NONE",
	NULL
};

/*
 * Names of special power flags.
 */
static char *effect_name[9][32] =
{
	/* Category zero doesn't exist */
	{
		NULL,
	},

	/* Category one */
	{
		"IGNORE",
		"INCREASE",
		"ONE_CHAR",
		"ALL_CHAR",
		"ONE_SUPPORT",
		"ALL_SUPPORT",
		"ONE_BOOSTER",
		"ALL_BOOSTER",
		"CATERPILLAR",
		"WITH_ICONS",
		"LEADERSHIP",
		"BLUFF",
		"ALL_CARDS",
		"TOTAL_POWER",
		"TOTAL_FIRE",
		"TOTAL_EARTH",
		"FIRE_VAL",
		"EARTH_VAL",
		"ODD_VAL",
		"EVEN_VAL",
		"SPECIAL",
		"ICONS_ALL",
		"ICONS_BUT_SP",
		"ICONS_BUT_S",
		"BY_FACTOR",
		"TO_VALUE",
		"BY_VALUE",
		"TO_SUM",
		"TO_HIGHER",
		"EXCEPT_FLIT",
		NULL,
	},

	/* Category two */
	{
		"I_RETREAT",
		"YOU_RETREAT",
		"ADDITIONAL",
		"FEWER",
		"EXACTLY",
		"NO_MORE_THAN",
		NULL,
	},


	/* Category three */
	{
		"YOU_MAY_NOT",
		"I_MAY_PLAY",
		"ADDITIONAL",
		"MORE_THAN",
		"DRAW",
		"TAKE",
		"CALL_BLUFF",
		"CHARACTER",
		"SUPPORT",
		"BOOSTER",
		"LEADERSHIP",
		"COMBAT",
		"HAVE_SPECIAL",
		"NO_SPECIAL",
		"WITH_VALUE",
		"AS_FREE",
		"SHIP_HAND",
		NULL,
	},
	
	/* Category four */
	{
		"DRAW",
		"DISCARD",
		"RETRIEVE",
		"REVEAL",
		"SEARCH",
		"UNDRAW_2",
		"SHUFFLE",
		"LOAD",
		"YOUR_HAND",
		"YOUR_CHAR",
		"YOUR_BOOSTER",
		"YOUR_SUPPORT",
		"YOUR_DECK",
		"MY_CHAR",
		"MY_BOOSTER",
		"MY_SUPPORT",
		"MY_HAND",
		"MY_DISCARD",
		"NOT_LAST_CHAR",
		"WITH_ICON",
		"ACTIVE",
		"DISCARD_ONE",
		"RANDOM_DISCARD",
		"TO",
		"ATTACK_AGAIN",
		"ON_BOTTOM",
		"OPTIONAL",
		"IF_FROM_SHIP",
		"ALL",
		"EITHER",
		NULL,
	},

	/* Category five */
	{
		"PLAY_ONLY_IF",
		"PLAY_FREE_IF",
		"FIRE_POWER",
		"EARTH_POWER",
		"EITHER_POWER",
		"YOU_ACTIVE",
		"YOU_PLAYED",
		"MY_PLAYED",
		"MY_INFLUENCE",
		"YOU_CHARACTER",
		"YOU_BOOSTER",
		"YOU_SUPPORT",
		"YOU_ICONS",
		"YOU_DRAGONS",
		"YOU_HANDSIZE",
		"ELEMENT_SWAP",
		NULL,
	},

	/* Category six */
	{
		"DISCARD",
		"STORM",
		"FIRE_VALUE",
		"EARTH_VALUE",
		"CHAR",
		NULL,
	},

	/* Category seven */
	{
		"PLAY_SUPPORT",
		"PLAY_BOOSTER",
		"DISCARD_FIRE",
		"DISCARD_EARTH",
		"DISCARD_BOTH",
		"DISCARD_EITHER",
		"DISCARD_CHAR",
		"CATERPILLAR",
		"FLOOD",
		"OR_RETREAT",
		"OR_DRAGON",
		NULL,
	},

	/* Category eight */
	{
		"YOU_DISCARD",
		"YOU_DISCLOSE",
		"TO",
		"OPTIONAL",
		NULL,
	},
};

/*
 * Lookup an effect code.
 */
static int lookup_effect(char *ptr, int cat)
{
	int i = 0;

	/* Loop over effects */
	while (effect_name[cat][i])
	{
		/* Check this effect */
		if (!strcmp(effect_name[cat][i], ptr)) return 1 << i;

		/* Next effect */
		i++;
	}

	/* No effect found */
	printf(_("No effect matching '%s'!\n"), ptr);
	exit(1);
}

/*
 * Read card designs from a text file.
 */
void read_cards(void)
{
	FILE *fff;
	char buf[1024], *ptr;
	int num_people = 0, num_design = 0;
	design *deck = NULL, *d_ptr = NULL;
	int i, effect, cat;

	/* Open card design file */
	fff = fopen(DATADIR "/cards.txt", "r");

	/* Check for error */
	if (!fff)
	{
		/* Try reading from current directory instead */
		fff = fopen("cards.txt", "r");
	}

	/* Check for error */
	if (!fff)
	{
		/* Print error and exit */
		perror("cards.txt");
		exit(1);
	}

	/* Loop over entire file */
	while (1)
	{
		/* Read a line */
		fgets(buf, 1024, fff);

		/* Check for end of file */
		if (feof(fff)) break;

		/* Strip newline */
		buf[strlen(buf) - 1] = '\0';

		/* Skip comments and blank lines */
		if (!buf[0] || buf[0] == '#') continue;

		/* Switch of type of line */
		switch (buf[0])
		{
			/* New people */
			case 'P':

				/* Read name */
				peoples[num_people].name = strdup(buf + 2);

				/* Pointer to current deck */
				deck = peoples[num_people].deck;

				/* Start at first design */
				num_design = 0;

				/* Count peoples */
				num_people++;

				break;

			/* New card */
			case 'N':

				/* Current design pointer */
				d_ptr = &deck[num_design];

				/* Read name */
				d_ptr->name = strdup(buf + 2);

				/* Count designs */
				num_design++;

				break;

			/* Strength values */
			case 'V':

				/* Get first value string */
				ptr = strtok(buf + 2, ":");

				/* Read fire value */
				d_ptr->value[0] = strtol(ptr, NULL, 0);

				/* Get next value */
				ptr = strtok(NULL, ":");

				/* Read earth value */
				d_ptr->value[1] = strtol(ptr, NULL, 0);

				break;

			/* Card type */
			case 'T':

				/* Read type */
				d_ptr->type = strtol(buf + 2, NULL, 0);

				break;

			/* Flags */
			case 'F':

				/* Get first flag */
				ptr = strtok(buf + 2, " |");

				/* Loop over flags */
				while (ptr)
				{
					/* Check each icon type */
					for (i = 0; icon_name[i]; i++)
					{
						/* Check this type */
						if (!strcmp(ptr, icon_name[i]))
						{
							/* Set icon flag */
							d_ptr->icons |= 1 << i;

							/* Done looking */
							break;
						}
					}

					/* Check for no match */
					if (!icon_name[i])
					{
						/* Error */
						printf("Unknown icon %s!\n",
						       ptr);

						/* Exit */
						exit(1);
					}

					/* Get next flag */
					ptr = strtok(NULL, " |");
				}

				break;

			/* Special text */
			case 'S':

				/* Read text */
				d_ptr->text = strdup(buf + 2);

				break;

			/* Special effect */
			case 'E':

				/* Get effect category string */
				ptr = strtok(buf + 2, ":");

				/* Read effect category */
				cat = strtol(ptr, NULL, 0);

				/* Store category */
				d_ptr->special_cat = cat;

				/* Get priority string */
				ptr = strtok(NULL, ":");

				/* Read priority */
				d_ptr->special_prio = strtol(ptr, NULL, 0);

				/* Get timing string */
				ptr = strtok(NULL, ":");

				/* Read timing */
				d_ptr->special_time = strtol(ptr, NULL, 0);

				/* Clear effect code */
				effect = 0;

				/* Read effect flags */
				while ((ptr = strtok(NULL, "|: ")))
				{
					/* Check for end of flags */
					if (isdigit(*ptr)) break;

					/* Lookup effect code */
					effect |= lookup_effect(ptr, cat);
				}

				/* Store effect code */
				d_ptr->special_effect = effect;

				/* Read effect value */
				d_ptr->special_value = strtol(ptr, NULL, 0);

				break;

			/* Miscellaneous info */
			case 'M':

				/* Get people string */
				ptr = strtok(buf + 2, ":");

				/* Read people */
				d_ptr->people = strtol(ptr, NULL, 0);

				/* Get index string */
				ptr = strtok(NULL, ":");

				/* Read index */
				d_ptr->index = strtol(ptr, NULL, 0);

				/* Get moon string */
				ptr = strtok(NULL, ":");

				/* Read moons */
				d_ptr->moons = strtol(ptr, NULL, 0);

				break;

			/* Capacity (for ships) */
			case 'C':

				/* Read capacity */
				d_ptr->capacity = strtol(buf + 2, NULL, 0);

				break;
		}
	}

	/* Close card design file */
	fclose(fff);
}

/*
 * Initialize a game.
 */
void init_game(game *g, int first)
{
	player *p;
	design *d_ptr;
	card *c;
	int i, j;

	/* Store game start random seed */
	g->start_seed = g->random_seed;

	/* printf("start seed: %u\n", g->start_seed); */

	/* Game not over yet */
	g->game_over = 0;

	/* Game is not a simulation */
	g->simulation = 0;

	/* No fight started or element chosen yet */
	g->fight_element = g->fight_started = 0;

	/* Loop over players */
	for (i = 0; i < 2; i++)
	{
		/* Get player pointer */
		p = &g->p[i];

		/* Loop over stacks */
		for (j = 0; j < LOC_MAX; j++) p->stack[j] = 0;

		/* Player has no dragons */
		p->dragons = 0;

		/* Player did not win instant victory */
		p->instant_win = 0;

		/* Player didn't run out of cards */
		p->no_cards = 0;

		/* Player may have no crystals */
		if (first) p->crystals = 0;

		/* Clear minimum power */
		p->min_power = 0;

		/* Get leader card */
		c = &p->deck[0];

		/* Set card design */
		c->d_ptr = &p->p_ptr->deck[0];

		/* Put in leadership pile */
		c->where = LOC_LEADERSHIP;

		/* Card not recently played */
		c->recent = 0;

		/* Remember last leadership card played */
		p->last_leader = c->d_ptr;

		/* No last discard yet */
		p->last_discard = NULL;

		/* Size of draw deck */
		p->stack[LOC_DRAW] = DECK_SIZE - 1;

		/* Loop over cards */
		for (j = 1; j < DECK_SIZE; j++)
		{
			/* Get card pointer */
			c = &p->deck[j];

			/* Set owner */
			c->owner = i;

			/* Set card design */
			c->d_ptr = &p->p_ptr->deck[j];

			/* Set card type */
			c->type = c->d_ptr->type;

			/* No target */
			c->target = NULL;

			/* Put card in draw pile */
			c->where = LOC_DRAW;

			/* Not on a ship */
			c->ship = NULL;

			/* Card is not stuck at bottom of draw pile */
			c->on_bottom = 0;

			/* Card is not a landed ship */
			c->landed = 0;

			/* Cards begin inactive */
			c->active = 0;

			/* Card is not played as free */
			c->was_played_free = 0;
			c->playing_free = 0;

			/* Card is not a bluff */
			c->bluff = 0;

			/* Card was not recently played */
			c->recent = 0;

			/* Cards are not fake */
			c->random_fake = 0;

			/* Card locations are unknown */
			c->loc_known = 0;
			c->disclosed = 0;
		}

		/* Draw six cards */
		for (j = 0; j < 6; j++)
		{
			/* Pick random card */
			d_ptr = random_card(g, i, LOC_DRAW);

			/* Move to hand */
			move_card(g, i, d_ptr, LOC_HAND, 0);
		}
	}

	/* Pick a starting player (randomly) */
	g->turn = myrand(&g->random_seed) % 2;

	/* Have start player begin */
	g->p[g->turn].phase = PHASE_START;
	g->p[!g->turn].phase = PHASE_NONE;

	/* Reset card flags */
	reset_cards(g);
}


learner.c
/*
 * Bluemoon AI
 * 
 * Copyright (C) 2007-2008 Keldon Jones
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include "bluemoon.h"

#include <signal.h>

/*
 * Be noisy?
 */
int verbose;

/*
 * No need for messages.
 */
void message_add(char *msg)
{
	/* Print messages if asked */
	if (verbose) printf("%s", msg);
}

/*
 * Initialize game and have AIs take all actions.
 */
int main(int argc, char *argv[])
{
	game my_game;
	player *p;
	int i, j, n = 100;

	/* Initialize random seed */
	my_game.random_seed = time(NULL);

	/* Read card designs */
	read_cards();

	/* Set default people */
	my_game.p[0].p_ptr = &peoples[0];
	my_game.p[1].p_ptr = &peoples[1];

	/* Parse arguments */
	for (i = 1; i < argc; i++)
	{
		/* Check for verbosity */
		if (!strcmp(argv[i], "-v"))
		{
			/* Set verbose flag */
			verbose += 1;
		}

		/* Check for number of games */
		else if (!strcmp(argv[i], "-n"))
		{
			/* Set number of games */
			n = atoi(argv[++i]);
		}

		/* Check for people setting */
		else if (!strcmp(argv[i], "-1") || !strcmp(argv[i], "-2"))
		{
			/* Loop over people */
			for (j = 0; j < MAX_PEOPLE; j++)
			{
				/* Check for match */
				if (!strcasecmp(argv[i + 1], peoples[j].name))
				{
					/* Set people */
					my_game.p[argv[i][1] - '1'].p_ptr =
					                           &peoples[j];
				}
			}

			/* Advance argument count */
			i++;
		}

		/* Check for random seed setting */
		else if (!strcmp(argv[i], "-r"))
		{
			/* Set random seed */
			my_game.random_seed = atoi(argv[i + 1]);

			/* Advance argument count */
			i++;
		}
	}

	/* Initialize game */
	init_game(&my_game, 1);

	/* Set player interface functions */
	my_game.p[0].control = &ai_func;
	my_game.p[1].control = &ai_func;

	/* Call interface initialization */
	for (i = 0; i < 2; i++)
	{
		/* Call init function */
		my_game.p[i].control->init(&my_game, i);
	}

	/* Play a number of games */
	for (i = 0; i < n; i++)
	{
		/* Take actions until game is over */
		while (1)
		{
			/* Get current player */
			p = &my_game.p[my_game.turn];

			/* Have AI take action */
			p->control->take_action(&my_game);

			/* Check for end of game */
			if (my_game.game_over)
			{
				/* Quit playing */
				break;
			}
		}

		/* Call game over functions */
		my_game.p[0].control->game_over(&my_game, 0);
		my_game.p[1].control->game_over(&my_game, 1);

		/* Message */
		printf("Crystals: %d %d\n", my_game.p[0].crystals,
		                            my_game.p[1].crystals);

		/* Restart game */
		init_game(&my_game, 1);
	}

	/* Call interface shutdown */
	for (i = 0; i < 2; i++)
	{
		/* Call shutdown function */
		my_game.p[i].control->shutdown(&my_game, i);
	}

	/* Done */
	return 0;
}


condition.c
/*
 * Bluemoon AI
 * 
 * Copyright (C) 2007-2008 Keldon Jones
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include "net.h"

/*
 * This extremely ugly code does some initial quick training of the
 * given neural net file.
 *
 * It sets only the "dragon" and "no cards" inputs and trains the
 * network about those.  Otherwise, the first several tens or hundreds
 * of thousands of training games are mostly wasted while the networks
 * learn about the scoring and victory conditions.
 */
int main(int argc, char *argv[])
{
	net learner;
	int i, n;
	double target[2];

	srand(time(NULL));

	make_learner(&learner, 443, 50, 2);

	learner.alpha = 0.1;

	load_net(&learner, argv[1]);

	for (i = 0; i < 443; i++) learner.input_value[i] = 0;

	/* Set "game over" input */
	learner.input_value[271] = 1;

	/* Do several training iterations */
	for (n = 0; n < 500; n++)
	{
		/* Set player 0 "ran out of cards first" input */
		learner.input_value[354] = 1;

		/* Set target scores */
		target[0] = 0.4;
		target[1] = 0.6;

		/* Train network */
		compute_net(&learner);
		train_net(&learner, 1.0, target);

		/* Clear no cards input */
		learner.input_value[354] = 0;

		/* Train dragon inputs */
		for (i = 0; i < 3; i++)
		{
			/* Set input */
			learner.input_value[355 + i] = 1;

			/* Compute desired outputs */
			target[0] = 0.5 + (i + 2) * 0.1;
			target[1] = 1.0 - target[0];

			/* Train network */
			compute_net(&learner);
			train_net(&learner, 1.0, target);
		}

		/* Set player 0 instant win input */
		learner.input_value[358] = 1;

		/* Compute desired outputs */
		target[0] = 1.0;
		target[1] = 0.0;

		/* Train network */
		compute_net(&learner);
		train_net(&learner, 1.0, target);

		/* Clear player 0 dragon inputs */
		for (i = 0; i < 3; i++) learner.input_value[355 + i] = 0;

		/* Clear player 0 instant win input */
		learner.input_value[358] = 0;

		/* Set player 1 "ran out of cards first" input */
		learner.input_value[438] = 1;

		/* Set target scores */
		target[0] = 0.6;
		target[1] = 0.4;

		/* Train network */
		compute_net(&learner);
		train_net(&learner, 1.0, target);

		/* Clear no cards input */
		learner.input_value[438] = 0;

		/* Train dragon inputs */
		for (i = 0; i < 3; i++)
		{
			/* Set input */
			learner.input_value[439 + i] = 1;

			/* Compute desired outputs */
			target[0] = 0.5 - (i + 2) * 0.1;
			target[1] = 1.0 - target[0];

			/* Train network */
			compute_net(&learner);
			train_net(&learner, 1.0, target);
		}

		/* Set player 1 instant win input */
		learner.input_value[442] = 1;

		/* Compute desired outputs */
		target[0] = 0.0;
		target[1] = 1.0;

		/* Train network */
		compute_net(&learner);
		train_net(&learner, 1.0, target);

		/* Clear player 1 dragon inputs */
		for (i = 0; i < 3; i++) learner.input_value[439 + i] = 0;

		/* Clear player 1 instant win input */
		learner.input_value[442] = 0;
	}

	save_net(&learner, argv[1]);
}


dumpnet.c
#include "bluemoon.h"
#include "net.h"

people peep[2];
int who;

void message_add(char *msg)
{
}

static void input_name(char *ptr, int i)
{
	int index, num;
	char *type;

	if (i < 240)
	{
		strcpy(ptr, peep[i / 120].name);

		strcat(ptr, " ");

		strcat(ptr, peep[i / 120].deck[i % 30 + 1].name);

		if (i % 120 < 30) strcat(ptr, " active");
		else if (i % 120 < 60) strcat(ptr, " in hand");
		else if (i % 120 < 90) strcat(ptr, " discarded");
		else strcat(ptr, " loaded");
	}
	else if (i < 270)
	{
		strcpy(ptr, peep[who].name);

		strcat(ptr, " ");

		strcat(ptr, peep[who].deck[i - 240 + 1].name);

		strcat(ptr, " special");
	}
	else if (i == 270) strcpy(ptr, "Bad bluff");
	else if (i == 271) strcpy(ptr, "Game over");
	else if (i == 272) strcpy(ptr, "Fight started");
	else if (i == 273) strcpy(ptr, "Fight in Earth");
	else if (i == 274) strcpy(ptr, "Fight in Fire");
	else
	{
		index = (i - 275) % 84;

		if (index == 0)
		{
			sprintf(ptr, "%s turn", peep[(i - 275) / 84].name);
			return;
		}
		index--;

		if (index == 78)
		{
			sprintf(ptr, "%s no cards", peep[(i - 275) / 84].name);
			return;
		}

		if (index == 82)
		{
			sprintf(ptr, "%s instant win", peep[(i - 275) / 84].name);
			return;
		}

		if (index < 15) type = "power";
		else if (index < 23) type = "combat";
		else if (index < 27) type = "bluff";
		else if (index < 37) type = "handsize";
		else if (index < 67) type = "decksize";
		else if (index < 72) type = "chars";
		else if (index < 78) type = "disclosed";
		else type = "dragons";

		if (index < 15) num = index;
		else if (index < 23) num = index - 15;
		else if (index < 27) num = index - 23;
		else if (index < 37) num = index - 27;
		else if (index < 67) num = index - 37;
		else if (index < 72) num = index - 67;
		else if (index < 78) num = index - 72;
		else num = index - 79;

		sprintf(ptr, "%s %s %d", peep[(i - 275) / 84].name, type, num);
	}
}

int main(int argc, char *argv[])
{
	net learner;
	int i;
	double start;
	char buf[1024], *ptr;
	int n = 0;

	read_cards();

	make_learner(&learner, 443, 50, 2);

	load_net(&learner, argv[1]);

	for (i = 0; i < 9; i++) if (strstr(argv[1], peoples[i].name)) peep[n++] = peoples[i];

	for (i = 0; i < 443; i++) learner.input_value[i] = 0;

	compute_net(&learner);

	who = atoi(argv[2]);

	start = learner.win_prob[who];

	for (i = 0; i < 443; i++)
	{
		learner.input_value[i] = 1;

		compute_net(&learner);

		input_name(buf, i);

		for (ptr = buf; *ptr; ptr++) if (*ptr == ' ') *ptr = '_';

		printf("%s: %f\n", buf, learner.win_prob[who] - start);

		learner.input_value[i] = 0;
	}
}
